librdmacm/cma.c

```diff
@@ -2595,7 +2605,9 @@ int rdma_get_cm_event(struct rdma_event_channel *channel,
 		evt->event.id = &evt->id_priv->id;
 		ucma_copy_ud_event(evt, &resp.param.ud);
 		evt->event.param.ud.private_data = evt->mc->context;
+		pthread_mutex_lock(&evt->id_priv->mut);
```

Comment:
I'm not sure that putting lock around ucma_process_join() is the right thing to do.

And more general comment, you want to make sure that id.qp pointer is valid, but didn't put the lock over all places where pointer to QP is used.

Reply:
In my opinion, what `ucma_process_join()` does is to invoke `ibv_attach_mcast()` on the id.qp if it exists, which may race with the concurrent call to `rdma_destroy_qp()`.

For functions other than `rdma_get_cm_event()` (e.g. `rdma_connect()`), it is trivial for users to implement a synchronization method **outside** the rdmacm by acquiring a per-cm_id lock before calling these functions. However, `rdma_get_cm_event()` is a different one, since multiple cm_ids may be attached to the same event channel. If id.qp can be accessed inside `rdma_get_cm_event()` without any synchronization, users may have to acquire a lock before calling these two functions, which is a per-event_channel lock and has higher contention.

Besides, users may choose to schedule all the `rdma_destroy_qp()` operations to the thread calling `rdma_get_cm_event()` to work around this issue, which effectively serialize these operations. But as it is noted in the manual: "Users must destroy any QP associated with an rdma_cm_id before destroying the ID", `rdma_destroy_id()` must be serialized with these functions as well if a QP is ever created on this cm_id, which is the common case in most applications. This makes the synchronization between `rdma_destroy_id()` and `rdma_get_cm_event()` less useful, which is already implemented inside rdmacm through mutex + condition variable and guarantees that "Destruction of an rdma_cm_id will block until related events have been acknowledged".

This MR mainly focuses on the issue that can not be easily addressed outside the rdmacm, so I choose not to put the lock where users can implement a per-cm_id lock as a workaround for the sake of performance. However, as you said, maybe we should put more locks for the sake of correctness.